wgengine/magicsock/magicsock.go

```diff
@@ -2599,86 +2595,91 @@ func (c *Conn) initialBind() error {
 	return nil
 }
 
-func (c *Conn) listenPacket(ctx context.Context, network, host string, port uint16) (net.PacketConn, error) {
+// listenPacket opens a listener. Host must be "" or an IP address.
+func (c *Conn) listenPacket(network, host string, port uint16) (net.PacketConn, error) {
+	ctx := context.Background() // unused without DNS name to resolve
 	addr := net.JoinHostPort(host, fmt.Sprint(port))
 	if c.packetListener != nil {
 		return c.packetListener.ListenPacket(ctx, network, addr)
 	}
 	return netns.Listener().ListenPacket(ctx, network, addr)
 }
 
-func (c *Conn) bind1(ruc **RebindingUDPConn, which string) error {
+// bind1 binds a UDP socket.
+// If there was an existing UDP socket bound, it closes it.
+// The caller is responsible for informing the portMapper of any changes.
+func (c *Conn) bind1(rucPtr **RebindingUDPConn, which string) error {
 	host := ""
 	if inTest() && !c.simulatedNetwork {
 		host = "127.0.0.1"
 		if which == "udp6" {
 			host = "::1"
 		}
 	}
-	var pc net.PacketConn
-	var err error
-	listenCtx := context.Background() // unused without DNS name to resolve
-	if c.port == 0 && DefaultPort != 0 {
-		pc, err = c.listenPacket(listenCtx, which, host, DefaultPort)
-		if err != nil {
-			c.logf("magicsock: bind: default port %s/%v unavailable; picking random", which, DefaultPort)
-		}
+
+	if *rucPtr == nil {
+		*rucPtr = new(RebindingUDPConn)
 	}
-	if pc == nil {
-		pc, err = c.listenPacket(listenCtx, which, host, c.port)
+	ruc := *rucPtr
+
+	// Hold the ruc lock the entire time, so that the close+bind is atomic
+	// from the perspective of ruc receive functions.
+	ruc.mu.Lock()
+	defer ruc.mu.Unlock()
+
+	// Build a list of preferred ports.
+	// Best is the port that the user requested.
+	// Second best is the port that is currently in use.
+	// If those fail, fall back to 0.
+	// The list of preferred ports might contain duplicates.
+	// Slightly wasteful, but no real harm to that.
```

Comment:
don't want to use your uniq.ModifySlice because it's there? (I don't care either way)

Reply:
Ooh, yeah, thanks.